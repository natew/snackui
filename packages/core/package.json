{
  "name": "@tamagui/core",
  "version": "1.2.7",
  "source": "src/index.ts",
  "main": "dist/cjs",
  "module": "dist/esm",
  "types": "./types/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Nate Wienert"
  },
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "tamagui-build",
    "watch": "tamagui-build --watch",
    "lint": "../../node_modules/.bin/rome check src",
    "lint:fix": "../../node_modules/.bin/rome check --apply-suggested src",
    "clean": "tamagui-build clean",
    "clean:build": "tamagui-build clean:build"
  },
  "files": [
    "types",
    "dist",
    "src",
    "reset.css"
  ],
  "dependencies": {
    "@tamagui/react-native-use-pressable": "^1.2.7",
    "@tamagui/react-native-use-responder-events": "^1.2.7",
    "@tamagui/web": "^1.2.7"
  },
  "peerDependencies": {
    "react": "*",
    "react-dom": "*"
  },
  "devDependencies": {
    "@tamagui/build": "^1.2.7",
    "@testing-library/react": "^13.4.0",
    "csstype": "^3.0.10",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "typescript": "^4.7.4",
    "vitest": "^0.26.3"
  },
  "exports": {
    "./package.json": "./package.json",
    "./reset.css": "./reset.css",
    ".": {
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./inject-styles": {
      "types": "./types/inject-styles.d.ts",
      "import": "./dist/esm/inject-styles.js",
      "require": "./dist/cjs/inject-styles.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/tamagui/tamagui.git",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "a49cc7ea6b93ba384e77a4880ae48ac4a5635c14"
}
