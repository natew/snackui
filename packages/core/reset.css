/* mimic the native viewport full width/height + flex */
body {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  min-width: 100vw;
}

* {
  box-sizing: border-box;
}

a {
  color: inherit;
}

p,
input,
textarea,
select,
button,
ul,
ol,
li,
pre,
h1, h2, h3, h4, h5, h6 {
  margin: 0;
  padding: 0;
}

li, select, input, button, [role="slider"] {
  outline: none;
}
