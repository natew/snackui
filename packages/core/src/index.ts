import { useResponderEvents } from '@tamagui/react-native-use-responder-events'
import {
  StackProps,
  StackPropsBase,
  TamaguiComponent,
  TextProps,
  TextPropsBase,
  Stack as WebStack,
  Text as WebText,
  composeEventHandlers,
  isRSC,
  mergeEvent,
  setupHooks,
  styled,
} from '@tamagui/web'
import { RefObject } from 'react'
import type { Text as RNText, View as RNView } from 'react-native'

import { useElementLayout } from './hooks/useElementLayout'
import { usePlatformMethods } from './hooks/usePlatformMethods'
import type { RNTextProps, RNViewProps } from './reactNativeTypes'
import { usePressability } from './vendor/Pressability'

// re-exports all of @tamagui/web just adds hooks
export * from '@tamagui/web'

// adds extra types to Stack/Text:

export const Stack = WebStack as TamaguiComponent<
  StackProps & RNViewProps,
  RNView,
  StackPropsBase & RNViewProps
>

export const Text = WebText as TamaguiComponent<
  TextProps & RNTextProps,
  RNText,
  TextPropsBase & RNTextProps
>

const T = styled(Stack, {
  backgroundColor: 'red',
  variants: {
    red: {
      true: {
        backgroundColor: 'red',
      },
    },
  },
})

const T2 = styled(T, {
  red: true,
})

// setup internal hooks:

setupHooks({
  getBaseViews() {
    const native = require('react-native')
    return {
      View: native.View || native.default.View,
      Text: native.Text || native.default.Text,
    }
  },

  usePropsTransform(elementType, propsIn, hostRef) {
    // otherwise replicate react-native-web functionality
    const {
      // event props
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onScrollShouldSetResponder,
      onScrollShouldSetResponderCapture,
      onSelectionChangeShouldSetResponder,
      onSelectionChangeShouldSetResponderCapture,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,

      // android
      collapsable,
      focusable,

      // deprecated,
      accessible,
      accessibilityDisabled,

      onLayout,
      hrefAttrs,

      ...viewProps
    } = propsIn

    if (!isRSC) {
      usePlatformMethods(hostRef as RefObject<Element>)
      useElementLayout(hostRef as RefObject<Element>, onLayout as any)
      useResponderEvents(hostRef, {
        onMoveShouldSetResponder,
        onMoveShouldSetResponderCapture,
        onResponderEnd,
        onResponderGrant,
        onResponderMove,
        onResponderReject,
        onResponderRelease,
        onResponderStart,
        onResponderTerminate,
        onResponderTerminationRequest,
        onScrollShouldSetResponder,
        onScrollShouldSetResponderCapture,
        onSelectionChangeShouldSetResponder,
        onSelectionChangeShouldSetResponderCapture,
        onStartShouldSetResponder,
        onStartShouldSetResponderCapture,
      } as any)
    }

    // TODO move into getSplitStyles inital `if (process.env.TAMAGUI_TARGET === 'web')` block

    if (viewProps.href !== undefined && hrefAttrs !== undefined) {
      const { download, rel, target } = hrefAttrs
      if (download != null) {
        viewProps.download = download
      }
      if (rel != null) {
        viewProps.rel = rel
      }
      if (typeof target === 'string') {
        viewProps.target = target.charAt(0) !== '_' ? `_${target}` : target
      }
    }

    // FOCUS
    // "focusable" indicates that an element may be a keyboard tab-stop.
    // ported from RNW: TODO move into getSplitStyles
    if (!viewProps.tabIndex) {
      const _focusable = focusable !== undefined ? focusable : accessible
      const role = viewProps.role
      if (_focusable === false) {
        viewProps.tabIndex = '-1'
      }
      if (
        // These native elements are focusable by default
        elementType === 'a' ||
        elementType === 'button' ||
        elementType === 'input' ||
        elementType === 'select' ||
        elementType === 'textarea'
      ) {
        if (_focusable === false || accessibilityDisabled === true) {
          viewProps.tabIndex = '-1'
        }
      } else if (
        // These roles are made focusable by default
        role === 'button' ||
        role === 'checkbox' ||
        role === 'link' ||
        role === 'radio' ||
        role === 'textbox' ||
        role === 'switch'
      ) {
        if (_focusable !== false) {
          viewProps.tabIndex = '0'
        }
      }
      // Everything else must explicitly set the prop
      if (_focusable === true) {
        viewProps.tabIndex = '0'
      }
    }

    return viewProps
  },

  useEvents(viewProps, events, { pseudos }, setStateShallow) {
    if (process.env.TAMAGUI_TARGET === 'native') {
      if (!events) return
      // add focus events
      const attachFocus = !!pseudos?.focusStyle
      if (attachFocus) {
        viewProps.onFocus = mergeEvent(viewProps.onFocus, () => {
          setStateShallow({ focus: true })
        })
        viewProps.onBlur = mergeEvent(viewProps.onBlur, () => {
          setStateShallow({ focus: false })
        })
      }
      // use Pressability to get smooth unPress when you press + hold + move out
      // only ever create once, use .configure() to update later
      const pressability = usePressability(
        events ? { ...events, hitSlop: viewProps.hitSlop } : {}
      )
      if (events.onPress) {
        for (const key in pressability) {
          const og = viewProps[key]
          const val = pressability[key]
          viewProps[key] =
            og && !dontComposePressabilityKeys[key] ? composeEventHandlers(og, val) : val
        }
      }
    }
  },
})

const dontComposePressabilityKeys = {
  onClick: true,
}
