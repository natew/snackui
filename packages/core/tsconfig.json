{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "composite": true,
    "strict": true,
    
  },
  "references": [
    {
      "path": "../helpers"
    },
    {
      "path": "../use-debounce"
    },
    {
      "path": "../use-force-update"
    },
  ]
}
