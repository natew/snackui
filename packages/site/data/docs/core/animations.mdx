---
title: Animations
description: Pluggable animation drivers.
---

<HeroContainer>
  <AnimationsDemo />
</HeroContainer>

```jsx hero template=Animations

```

<Highlights
  disableLinks
  features={[
    `Pluggable drivers, fully typed and easy to swap.`,
    `Support hover, press, and focus animations.`,
    `Supports enter animation on mount.`,
    `Supports exit animation with AnimatePresence.`,
  ]}
/>

Add animations to Tamagui with an animation driver.

Animation drivers are designed to be swappable, so you can use lightweight CSS animations or other web-focused animation libraries on the web, while using larger but more advanced libraries like `reanimated` on native - all without having to change a line outside of configuration.

## Installation

### CSS

The `@tamagui/animations-css` package works with the tamagui compiler and runtime to give you simple ways to share typed animations across all your components.

To install it add to your package.json:

```bash
yarn add @tamagui/animations-css
```

Then add it to your [config](/docs/intro/configuration):

```tsx
import { createAnimations } from '@tamagui/animations-css'
import { createTamagui } from 'tamagui'

export default createTamagui({
  animations: createAnimations({
    fast: 'ease-in 150ms',
    medium: 'ease-in 300ms',
    slow: 'ease-in 450ms',
  }),
  // ...
})
```

At runtime, the plugin does very little except to set the `transition` property in CSS. At compile-time, the compiler do the same, ensuring you get all the benefits of prop removal and view flattening even when using animations.

### Reanimated

[Reanimated](https://docs.swmansion.com/react-native-reanimated/) is an animation library that targets React Native and React Native Web. It runs off-thread animations, and provides nicer syntax and hooks. It weighs over 100Kb minified though, which makes it not always useful for the web.

To install it add to your package.json:

```bash
yarn add @tamagui/animations-reanimated
```

Then add it to your [config](/docs/intro/configuration):

```tsx
import { createAnimations } from '@tamagui/animations-reanimated'
import { createTamagui } from 'tamagui'

export default createTamagui({
  animations: createAnimations({
    fast: {
      type: 'spring',
      damping: 20,
      mass: 1.2,
      stiffness: 250,
    },
    medium: {
      type: 'spring',
      damping: 10,
      mass: 0.9,
      stiffness: 100,
    },
    slow: {
      type: 'spring',
      damping: 20,
      stiffness: 60,
    },
  }),
  // ...
})
```

At runtime, this plugin parses animatable style properties and hands them over to reanimated off-thread, using worklets. It doesn't do anything at compile time - reanimated must run via JS.

---

Note the keys match between CSS and reanimated, so you can swap them out.

<Notice>
  **Save bundle size** by removing `Animated` and `AnimatedFlatList` from react-native-web. On
  Next.js, use the [excludeReactNativeWebExports` option](/docs/intro/installation#next.js) config
  option.
</Notice>

### Usage

The `animation` can now accept `slow` as a value. By default, animations will apply to all animatable styles, a lot like setting `all` in a CSS `transition`. Let's test this by setting `hoverStyle`:

Here's an example animating `hoverStyle`:

<HeroContainer>
  <AnimationsHoverDemo />
</HeroContainer>

```jsx hero template=AnimationsHover

```

### enterStyle

Setting `enterStyle` styles on any component tell it to start with those styles, and immediately transition into their regular styles:

<HeroContainer>
  <AnimationsEnterDemo />
</HeroContainer>

```jsx hero template=AnimationsEnter

```

### AnimatePresence and exitStyle

AnimatePresence animates direct children before they unmount. It's inspired by and forked from [Framer Motion](https://github.com/framer/motion), but works with any animation in Tamagui.

To use with `@tamagui/core`, install and import `@tamagui/animate-presence`. It's already bundled and exported from `tamagui`.

```tsx
import { AnimatePresence, Square } from 'tamagui'

export const MyComponent = ({ isVisible }) => (
  <AnimatePresence>
    {isVisible && (
      <Square
        key="my-square"
        animation="bounce"
        backgroundColor="green"
        size={50}
        exitStyle={{
          size: 100,
        }}
      />
    )}
  </AnimatePresence>
)
```

Wrap one or more tamagui components with AnimatePresence. This enables the use the exitStyle prop, whichÂ accepts all style properties and variants, which will be animated to using the `animation` defined before the component is unmounted.

<Notice>
  Animated child components must each have a unique key prop so AnimatePresence can track their
  presence in the tree.
</Notice>

<HeroContainer noPad>
  <AnimationsPresenceDemo />
</HeroContainer>

```jsx hero template=AnimationsPresence

```

AnimatePresence takes two optional special properties, `enterVariant` and `exitVariant`, which allow you to override enterStyle and exitStyle conditionally. These properties would be impossible to change otherwise. This is a simple example, hit "Show Code" for the carousel demo above to see a more practical usage of it:

```tsx
import { AnimatePresence, Square, styled } from 'tamagui'

const AnimatableSquare = styled(Square, {
  variants: {
    fromRight: {
      true: {
        x: 1000,
      },
    },
    fromLeft: {
      true: {
        x: -1000,
      },
    },
  },
})

export const MyComponent = ({ isVisible }) => (
  <AnimatePresence enterVariant="fromLeft" exitVariant="fromRight">
    {isVisible && (
      <Square
        key="my-square"
        animation="bounce"
        backgroundColor="green"
        size={50}
        exitStyle={{
          opacity: 0,
        }}
      />
    )}
  </AnimatePresence>
)
```

Note that it will always find the `true` variant and use that. In the future, Tamagui may be able to add support for more flexible variants here.
