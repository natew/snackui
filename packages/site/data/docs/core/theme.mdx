---
title: Theme
description: Changing themes in your app.
---

<Notice>
  The package `tamagui` is a superset of `@tamagui/core`. If you are using `tamagui`, be sure to
  import Theme from there - there's no need to add core to your package.json.
</Notice>

Changing themes in Tamagui is as easy as passing their name to the Theme component.

### Usage

Change themes anywhere in your app like so:

```tsx
import { Button, Theme } from '@tamagui/core'

export default () => (
  <Theme name="dark">
    <Button>I'm a dark button</Button>
  </Theme>
)
```

### Sub-themes

A good example of this setup is [the source of this site itself](https://github.com/tamagui/tamagui/blob/master/packages/site/constants/themes.ts).

There is one special property of themes that's very helpful for what is essentially sub-themes, or tints.

Basically, if you have a base `dark` theme and a `light` theme, you may want to have "tinted" versions of those themes. For example `light_pink` and `dark_pink`. Instead of having hard-coded color values like `$pink3`, you create as many pink themes as you want: `light_pink_alt1`, `light_pink_alt2`, etc.

Each of those themes is recommended to be small: just having values that map to ViewStyle or TextStyle, like `$color` (and that can postfix pseudo states, like `$colorHover`).

Then, you just give your component a `theme="pink"` prop, or wrap it in `<Theme name="pink" />`. Note you can leave out the `light_` or even `light_pink` prefix, Tamagaui will nest properly so long as the parent is set.

It's easier to understand with an example. First, define your tokens and themes:

```tsx
import { createTamagui, createTokens } from 'tamagui'

const tokens = createTokens({
  color: {
    pinkDark: '#610c62',
    pinkLight: '#f17efc',
  },
})

export default createTamagui({
  tokens,
  themes: {
    dark: {
      background: '#000',
      color: '#fff',
    },
    light: {
      color: '#000',
      background: '#fff',
    },
    dark_pink: {
      background: tokens.color.pinkDark,
      color: tokens.color.pinkLight,
    },
    light_pink: {
      background: tokens.color.pinkLight,
      color: tokens.color.pinkDark,
    },
  },
})
```

Using the `_pink` suffix, the `Theme` component will now let us automatically do subsetting:

```tsx
import { Button, Theme } from 'tamagui'

export default () => {
  return (
    <Theme name="dark">
      <Button>I have the theme dark</Button>
      <Theme name="pink">
        <Button>I have the theme pink-dark</Button>
      </Theme>
    </Theme>
  )
}
```

Notice we just use the name `pink`. Because theres no `pink` theme, the Theme component will automatically check if theres a theme that matches the pattern `[currentThemeName]_[givenName]`, in this case `dark_pink`. This is really useful for things like having an `active` or `error` theme that match your parent theme styles.

---

## Theme Inverse

The `ThemeInverse` component is handy for flipping from light to dark without having to do manual logic with context.

<HeroContainer>
  <ThemeInverseDemo />
</HeroContainer>

```tsx hero template=ThemeInverse

```

---

## themeable()

The `themeable` helper function is a "higher order component" that wraps any React component definition and adds two props:

- `theme` for changing the theme on that component and its children
- `themeInverse` which wraps the component with `<ThemeInverse>`

It's used for example on [Button](/docs/components/button).
