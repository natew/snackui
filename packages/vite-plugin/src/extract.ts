// fork from https://github.com/seek-oss/vanilla-extract

import path from 'path'

import { TamaguiOptions } from '@tamagui/static'
import * as Static from '@tamagui/static'
import outdent from 'outdent'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import { normalizePath } from 'vite'

const { createExtractor, extractToClassNames, getPragmaOptions } =
  Static['default'] || Static

const styleUpdateEvent = (fileId: string) => `tamagui-style-update:${fileId}`
const GLOBAL_CSS_VIRTUAL_PATH = '__tamagui_global_css__.css'

export function tamaguiExtractPlugin(options: TamaguiOptions): Plugin {
  const disableStatic =
    options.disable || (options.disableDebugAttr && options.disableExtraction)

  if (disableStatic) {
    return {
      name: 'tamagui-extract',
    }
  }

  let extractor: ReturnType<typeof createExtractor> | null = null
  const cssMap = new Map<string, string>()

  let config: ResolvedConfig
  let server: ViteDevServer
  let shouldReturnCSS = true //config.command === 'serve'
  let virtualExt: string

  const getAbsoluteVirtualFileId = (filePath: string) => {
    if (filePath.startsWith(config.root)) {
      return filePath
    }
    return normalizePath(path.join(config.root, filePath))
  }

  return {
    name: 'tamagui-extract',
    enforce: 'pre',

    configureServer(_server) {
      server = _server
    },

    buildEnd() {
      extractor!.cleanupBeforeExit()
    },

    writeBundle(this, options, bundle) {
      setTimeout(() => {
        // rome-ignore lint/nursery/noConsoleLog: ok
        console.warn('some sort of dangling process or osmethign, exit for now...')
        process.exit(0)
      }, 100)
    },

    config(_userConfig, env) {
      const include = env.command === 'serve' ? ['@tamagui/core/inject-styles'] : []
      return {
        optimizeDeps: { include },
      }
    },

    async configResolved(resolvedConfig) {
      config = resolvedConfig
      extractor = createExtractor({
        logger: resolvedConfig.logger,
      })
      shouldReturnCSS = true
      // TODO postcss work with postcss.config.js
      // packageName = getPackageInfo(config.root).name;
      // if (config.command === 'serve') {
      //   postCssConfig = await resolvePostcssConfig(config);
      // }
      virtualExt = `.tamagui.${shouldReturnCSS ? 'css' : 'js'}`
    },

    async resolveId(source) {
      if (source === 'tamagui.css') {
        await extractor!.loadTamagui(options)
        return GLOBAL_CSS_VIRTUAL_PATH
      }

      const [validId, query] = source.split('?')

      if (!validId.endsWith(virtualExt)) {
        return
      }

      // Absolute paths seem to occur often in monorepos, where files are
      // imported from outside the config root.
      const absoluteId = source.startsWith(config.root)
        ? source
        : getAbsoluteVirtualFileId(validId)

      // There should always be an entry in the `cssMap` here.
      // The only valid scenario for a missing one is if someone had written
      // a file in their app using the .tamagui.js/.tamagui.css extension
      if (cssMap.has(absoluteId)) {
        // Keep the original query string for HMR.
        return absoluteId + (query ? `?${query}` : '')
      }
    },

    /**
     * TODO
     *
     *   mainFields module:jsx breaks, so lets just have a mapping here
     *   where we load() and map it to the jsx path before transform
     *
     */

    load(id, options) {
      const [validId] = id.split('?')

      if (validId === GLOBAL_CSS_VIRTUAL_PATH) {
        return extractor!.getTamagui()!.getCSS()
      }

      if (!cssMap.has(validId)) {
        return
      }

      const css = cssMap.get(validId)

      if (typeof css !== 'string') {
        return
      }

      if (shouldReturnCSS || !server || server.config.isProduction) {
        return css
      }

      return outdent`
        import { injectStyles } from '@tamagui/core/inject-styles';

        const inject = (css) => injectStyles({
          filePath: "${validId}",
          css
        });

        inject(${JSON.stringify(css)});

        if (import.meta.hot) {
          import.meta.hot.on('${styleUpdateEvent(validId)}', (css) => {
            inject(css);
          });
        }
      `
    },

    async transform(code, id, ssrParam) {
      const [validId] = id.split('?')

      if (!validId.endsWith('.tsx')) {
        return
      }

      // let ssr: boolean | undefined
      // if (typeof ssrParam === 'boolean') {
      //   ssr = ssrParam
      // } else {
      //   ssr = ssrParam?.ssr
      // }

      const firstCommentIndex = code.indexOf('// ')
      const { shouldDisable, shouldPrintDebug } = getPragmaOptions({
        source: firstCommentIndex >= 0 ? code.slice(firstCommentIndex) : '',
        path: validId,
      })

      if (shouldDisable) {
        return
      }

      const extracted = await extractToClassNames({
        extractor: extractor!,
        source: code,
        sourcePath: validId,
        options,
        shouldPrintDebug,
      })

      if (!extracted) {
        return
      }

      const rootRelativeId = `${validId}${virtualExt}`
      const absoluteId = getAbsoluteVirtualFileId(rootRelativeId)

      let source = extracted.js

      if (extracted.styles) {
        if (
          server &&
          cssMap.has(absoluteId) &&
          cssMap.get(absoluteId) !== extracted.styles
        ) {
          const { moduleGraph } = server
          const [module] = Array.from(moduleGraph.getModulesByFile(absoluteId) || [])

          if (module) {
            moduleGraph.invalidateModule(module)

            // Vite uses this timestamp to add `?t=` query string automatically for HMR.
            module.lastHMRTimestamp =
              (module as any).lastInvalidationTimestamp || Date.now()
          }

          server.ws.send({
            type: 'custom',
            event: styleUpdateEvent(absoluteId),
            data: extracted.styles,
          })
        }

        source = `${source}\nimport "${rootRelativeId}";`
        cssMap.set(absoluteId, extracted.styles)
      }

      return {
        code: source.toString(),
        map: extracted.map,
      }

      // if (ssr && !process.env.VITE_RSC_BUILD) {
      //   return addFileScope({
      //     source: code,
      //     filePath: normalizePath(validId),
      //     rootPath: config.root,
      //     packageName,
      //   })
      // }

      // const { source, watchFiles } = await compile({
      //   filePath: validId,
      //   cwd: config.root,
      //   esbuildOptions,
      // })

      // for (const file of watchFiles) {
      //   // In start mode, we need to prevent the file from rewatching itself.
      //   // If it's a `build --watch`, it needs to watch everything.
      //   if (config.command === 'build' || file !== validId) {
      //     this.addWatchFile(file)
      //   }
      // }
    },
  }
}
