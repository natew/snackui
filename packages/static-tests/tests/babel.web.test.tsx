import * as React from 'react'
import { expect, test } from 'vitest'

import { extractForNative, extractForWeb } from './lib/extract'

process.env.IS_STATIC = ''

window['React'] = React

test('conditional styles get full base styles merged onto + shorthand', async () => {
  console.log('RUN')

  // one sanity check debug output test
  const output = await extractForWeb(
    `
import { Stack } from '@tamagui/core'
    export function Test(props) {
      return (
        <Stack width={10} bg={props.green ? 'red' : 'blue'} />
      )
    }
  `,
    {
      options: {
        components: ['@tamagui/core'],
      },
    }
  )

  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})

test('className + conditional styles get full base styles merged onto + shorthand', async () => {
  // one sanity check debug output test
  const output = await extractForWeb(
    `
    import { Stack } from '@tamagui/core'
    export function Test(props) {
      return (
        <Stack width={10} bg={props.green ? 'red' : 'blue'} className={props.className} />
      )
    }
  `,
    {
      options: {
        components: ['@tamagui/core'],
      },
    }
  )
  expect(output?.js).toMatchSnapshot()
  expect(output?.styles).toMatchSnapshot()
})
