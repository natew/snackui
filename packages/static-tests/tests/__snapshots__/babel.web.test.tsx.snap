// Vitest Snapshot v1

exports[`className + conditional styles get full base styles merged onto + shorthand 1`] = `
"const _cn3 = \\" _bg-d0t1494150182\\";
const _cn2 = \\" _bg-d0t1948734170\\";
const _cn = \\" _w-10px _fd-column _miw-0px _mih-0px _pos-relative _bxs-border-box _fb-auto _dsp-flex _fs-0 _ai-stretch  \\";
import { Stack } from '@tamagui/core';
export function Test(props) {
  return <div className={concatClassName(\\"\\" + (_cn + (props.green ? _cn2 : _cn3) + (props.className ? \\" \\" + props.className : \\"\\")))} />;
}
import { concatClassName } from \\"@tamagui/helpers\\";"
`;

exports[`className + conditional styles get full base styles merged onto + shorthand 2`] = `
":root ._w-10px{width:10px;}
:root ._fd-column{flex-direction:column;}
:root ._miw-0px{min-width:0px;}
:root ._mih-0px{min-height:0px;}
:root ._pos-relative{position:relative;}
:root ._bxs-border-box{box-sizing:border-box;}
:root ._fb-auto{flex-basis:auto;}
:root ._dsp-flex{display:flex;}
:root ._fs-0{flex-shrink:0;}
:root ._ai-stretch{align-items:stretch;}
:root ._bg-d0t1948734170{background-color:rgba(255,0,0,1.00);}
:root ._bg-d0t1494150182{background-color:rgba(0,0,255,1.00);}"
`;

exports[`conditional styles get full base styles merged onto + shorthand 1`] = `
"const _cn3 = \\" _bg-d0t1494150182\\";
const _cn2 = \\" _bg-d0t1948734170\\";
const _cn = \\" _w-10px _fd-column _miw-0px _mih-0px _pos-relative _bxs-border-box _fb-auto _dsp-flex _fs-0 _ai-stretch  \\";
import { Stack } from '@tamagui/core';
export function Test(props) {
  return <div className={concatClassName(\\"\\" + (_cn + (props.green ? _cn2 : _cn3)))} />;
}
import { concatClassName } from \\"@tamagui/helpers\\";"
`;

exports[`conditional styles get full base styles merged onto + shorthand 2`] = `
":root ._w-10px{width:10px;}
:root ._fd-column{flex-direction:column;}
:root ._miw-0px{min-width:0px;}
:root ._mih-0px{min-height:0px;}
:root ._pos-relative{position:relative;}
:root ._bxs-border-box{box-sizing:border-box;}
:root ._fb-auto{flex-basis:auto;}
:root ._dsp-flex{display:flex;}
:root ._fs-0{flex-shrink:0;}
:root ._ai-stretch{align-items:stretch;}
:root ._bg-d0t1948734170{background-color:rgba(255,0,0,1.00);}
:root ._bg-d0t1494150182{background-color:rgba(0,0,255,1.00);}"
`;
